test/e2e/multihoming_external_router_utils.go

```diff
@@ -0,0 +1,429 @@
+// and the gateway mode:
+// |        | Local GW                  | Shared GW                                                                  |
+// | IC     | linux route on all node   | linux routes on all nodes; OVN routes on all nodes for the local GW router |
+// | non-IC | linux routes on all nodes | linux routes on all nodes; OVN routes on NBDB leader for all GW routers    |
+func injectStaticRoutesIntoNodes(f *framework.Framework, cs clientset.Interface, externalContainerName string) error {
+	framework.Logf("Injecting Linux kernel routes for host-networked pods (and for OVN pods when in local gateway mode)")
+	if err := injectRoutesWithCommandBuilder(f, cs, externalContainerName, hostRoutingTableCommandBuilder{}); err != nil {
+		return err
+	}
+
+	if !IsGatewayModeLocal(cs) {
+		framework.Logf("Shared gateway mode: injecting OVN routes for overlay pods")
+		if err := injectRoutesWithCommandBuilder(f, cs, externalContainerName, ovnLogicalRouterCommandBuilder{}); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// routeCommand represents a route add/delete command for a specific gateway mode
+type routeCommand struct {
+	addCmd    []string
+	deleteCmd []string
+	logMsg    string
+	target    string // what we're adding the route to (logical router name or node name)
+}
+
+// routeCommandBuilder defines the interface for building gateway-mode-specific route commands
+type routeCommandBuilder interface {
+	buildRouteCommand(nodeName, cidr, nextHop string) routeCommand
+}
+
+// ovnLogicalRouterCommandBuilder builds commands for OVN logical routes (shared gateway mode)
+type ovnLogicalRouterCommandBuilder struct{}
+
+func (b ovnLogicalRouterCommandBuilder) buildRouteCommand(nodeName, cidr, nextHop string) routeCommand {
+	logicalRouterName := "GR_" + nodeName
+	return routeCommand{
+		addCmd:    []string{"ovn-nbctl", "--may-exist", "--", "lr-route-add", logicalRouterName, cidr, nextHop},
+		deleteCmd: []string{"ovn-nbctl", "--if-exists", "--", "lr-route-del", logicalRouterName, cidr},
+		logMsg:    fmt.Sprintf("OVN logical router route %s via %s to %s", cidr, nextHop, logicalRouterName),
+		target:    logicalRouterName,
+	}
+}
+
+// hostRoutingTableCommandBuilder builds commands for routes on the host
+type hostRoutingTableCommandBuilder struct{}
+
+func (b hostRoutingTableCommandBuilder) buildRouteCommand(nodeName, cidr, nextHop string) routeCommand {
+	return routeCommand{
+		addCmd:    []string{"ip", "route", "replace", cidr, "via", nextHop},
+		deleteCmd: []string{"ip", "route", "del", cidr, "via", nextHop},
+		logMsg:    fmt.Sprintf("host route %s via %s to node %s", cidr, nextHop, nodeName),
+		target:    nodeName,
+	}
+}
+
+// getOvnKubePodsForRouteInjection determines which pods to use for route injection based on the command builder type
+// and cluster configuration.
+func getOvnKubePodsForRouteInjection(f *framework.Framework, cs clientset.Interface, cmdBuilder routeCommandBuilder) (*v1.PodList, error) {
```

Comment:
this and some other utils here... are we sure we can't reuse existing utils from e2e.go file?
we have plumbing for some of this but maybe not exactly specific to your test here.

Reply:
yeah, I didn't find exactly what I needed.